python-lib/prologin/matchmaker/matchmaker.py

```diff
@@ -0,0 +1,308 @@
+
+        # Last champion of each user
+        # https://stackoverflow.com/questions/16074498
+        chs_ids = (
+            all_chs.values("author__id")
+            .annotate(max_id=Max("id"))
+            .values("max_id")
+        )
+
+        return Champion.objects.filter(pk__in=chs_ids)
+
+    async def setup(self):
+        self.author = django.contrib.auth.get_user_model().objects.get(
+            pk=self.config["matchmaker"]["author_id"]
+        )
+        logging.info("setup: match author: %s", self.author)
+
+        self.tournament, _ = Tournament.objects.get_or_create(
+            name=self.config["matchmaker"]["tournament_name"],
+            defaults={"author": self.author},
+        )
+        logging.info("setup: tournament: %s", self.tournament)
+
+        self.maps = list(self.tournament.maps.all())
+        logging.info("setup: tournament maps: %s", self.maps)
+
+    async def bootstrap(self):
+        # Get current champions
+        current_champions = self.get_champions_query()
+        self.champions = set(current_champions)
+        logging.info("bootstrap: champions: %d", len(self.champions))
+
+        # Remove old champions
+        old_champions = self.tournament.players.exclude(
+            pk__in=current_champions
+        ).all()
+        logging.info(
+            "bootstrap: removing %d old champion(s) from tournament",
+            len(old_champions),
+        )
+        self.tournament.players.remove(*old_champions)
+
+        # Add new champions
+        previous_champions = self.tournament.players.all()
+        new_champions = [
+            c for c in self.champions if c not in previous_champions
+        ]
+        logging.info(
+            "bootstrap: adding %d new champion(s) to the tournament",
+            len(new_champions),
+        )
+        self.tournament.players.add(*new_champions)
+
+        # Remove old matches
+        await self.cancel_matches(old_champions, self.champions)
+
+        # Create missing matches
+        logging.debug('bootstrap: computing missing matches')
+        desired_matches = Counter(self.get_desired_matches(self.champions))
+        current_matches = (
+            MatchItem.from_db(m)
+            for m in self.tournament.matches.all().prefetch_related('players')
+        )
+        desired_matches.subtract(current_matches)
+        missing_matches = list(desired_matches.elements())
+        logging.info(
+            'bootstrap: creating %d missing matches', len(missing_matches)
+        )
+        random.shuffle(missing_matches)
+        self.add_matches(missing_matches)
+
+    def get_desired_matches(self, champions):
+        """Yields all expected matches for champions."""
+        for chs in itertools.product(
+            champions, repeat=django_settings.STECHEC_NPLAYERS
+        ):
+            # Don't fight against yourself
+            ch_ids = [c.id for c in chs]
+            if len(set(ch_ids)) != len(ch_ids):
```

Comment:
We don't just want a N-combination of the champions.

What the current algorithm produces is both a combinations then a permutation without duplicate entries, so that each champion runs against all other champions, in all possible orders. Using the full product seems to be the simplest and clearest way to achieve that.

Reply:
Discussed offline → now using itertools.permutations().